### PR TITLE
Sliding window counter

### PR DESCRIPTION
**Pull request by Merinorus**
Following discussion https://github.com/alisaifee/limits/discussions/245

Sliding window counter implementation.
Supported backends: in-memory, redis, memcached.

TODO:
- [x] Floor the weighted counter to avoid premature rate limit. update documentation.
- [x] Memcached: implementation is sensitive to clock drift between limiter workers (time sync is mandatory)
- [ ] How to emulate slow IO (get, incr...) for any storage in tests?
- [ ] Memcached: why setting noreply=True leads to worse IOPS?
- [x] Memcached: remove expiry counter creation (not needed) to increase performance
- [x] In-memory (async): concurrency problem, can allow more requests than expected -> timestamp-based keys fix this issue
- [x] in-memory (sync): check if the window shift is ok against high concurrency / multi threaded environment. Otherwise use the timestamp based keys -> timestamp-based keys like the async version
- [x] Add tests: reset time memcached/other, weighted counter, concurrency?
- [x] compute reset time (different if memcached or other)
- [x] Remove test application, remove comments, clean code
- [x] Squash in one commit

### CODE REVIEW

**Comment by alisaifee** [comment]
Small tip: to fix linting issues locally before pushing you can run:
```
make lint-fix
```
and to just check:
```
make lint
```

(You'd ofcourse need all the development dependencies by running `pip install -r requirements/dev.txt`)

**Comment by Merinorus** [comment]
Thanks for the tip! I was doing the Ruff commands one by one, sometimes missing one.

We'll have to double-check the Memcached and Redis algorithms if they are valid on paper.
The performance on Redis and in memory seems on par with the moving window algorithm. The Memcached is a bit slower but, if it works, I will leave it as it is for now.

I have to check also the async in-memory storage concurrency, sometimes it allows one more hit than expected.
```
=================================== FAILURES ===================================
____________ TestConcurrency.test_sliding_window_counter[in-memory] ____________
tests/integration/test_concurrency.py:67: in test_sliding_window_counter
    assert len(hits) == 5
E   assert 6 == 5
E    +  where 6 = len([None, None, None, None, None, None])
=============================== warnings summary ===============================
```

What do you think about the current draft?

**Comment by alisaifee** [comment]
I haven't fully thought (or tested) this through, but just throwing it out there in case it makes sense.

For the current / previous key - could we use a different scheme where:
```py
now = int(time.time())
current_key = limit.key_for(*[*identifiers, str(int(now / item.get_expiry()))])
prev_key = limit.key_for(*[*identifiers, str(int((now - item.get_expiry()) / item.get_expiry()))])
```

And then use `item.get_expiry() * 2` for each key.

This way you wouldn't have to migrate the count from current to previous

**Comment by alisaifee** [comment]
Perhaps we could also push the onus of previous/current key into the storage so that the storage splits the key into two. This could have the additional benefit of only adding the `{` `}` brackets in the redis cluster implementation where it makes sense.

**Comment by Merinorus** [comment]
> I haven't fully thought (or tested) this through, but just throwing it out there in case it makes sense.
> 
> For the current / previous key - could we use a different scheme where:
> 
> ```python
> now = int(time.time())
> current_key = limit.key_for(*[*identifiers, str(int(now / item.get_expiry()))])
> prev_key = limit.key_for(*[*identifiers, str(int((now - item.get_expiry()) / item.get_expiry()))])
> ```
> 
> And then use `item.get_expiry() * 2` for each key.
> 
> This way you wouldn't have to migrate the count from current to previous

Yes indeed, it makes sense. I thought about this kind of solution for memcached.

Would you use this scheme on redis as well? What I like about the current solution is that it depends on the redis clock only, it's not mixed with the rate limit worker's clock. I'm not sure what is the convention, but I would be more confident using this clock, assuming redis cluster nodes have mechanisms to keep the clock in sync (and there should be no problem with a single redis instance, even if it's not synced).

(current work saved on https://github.com/Merinorus/limits/tree/2025-01-10-sliding-window-counter)

**Comment by alisaifee** [comment]
> Would you use this scheme on redis as well? What I like about the current solution is that it depends on the redis clock only, it's not mixed with the rate limit worker's clock. 

If we agree on moving the splitting of the key into the storage implementation (i.e. the strategy doesn't know about which clock is being used) then we can perhaps fine tune this depending on the capabilities of the storage.

**Comment by Merinorus** [comment]
Hi,

So I moved the splitting logic into the storage implementation.
For the Memcached implementation, I did the following:

1. previous and current keys are computed according to the current time and the expiry time,
2. key shifting and lock are removed _(no performance gain, but not having to lock anything is comforting?)_,
3. rate limiter hit attempts now need three Memcached variables instead of four _(almost no performance gain and complicated code, I intend to remove this part)_.

The limiter reset time is now more predictable, although this is no big deal.
However, the performance gain is negligible: the lock and time shift logic happened once per sampling period.

I did something else: for Memcached counter incrementation, a check was made at first with "add". I removed this check by directly doing counter incrementation. If I have a return value, the key already exists and I return the value. Otherwise, it doesn't exist yet and I re-run the exact counter logic written before. This makes the first incrementation slower, but all the subsequent writes are much faster. I have to double-check if it doesn't break anything.

With Memcached, the rate limiter median response time dropped from 1081 µs to 677 µs.
With the Memcached cluster, it dropped from 1274 µs to 1081 µs.
The fixed-window rate limiter with Memcached should be faster as well.

I think we can optimize the Memcached cluster response by having the current key and the previous key on the same node with some hashing logic.

Let me know! I'll rework the async part when we agree on this one.

**Comment by alisaifee** [comment]
> Let me know! I'll rework the async part when we agree on this one.

This looks great, in fact your changes to memcached is definitely a "more" correct approach (I couldn't reason why I'd chosen add first, but to be fair that was more than a decade ago 😁)

I'm happy to proceed with this approach. Once you're mostly comfortable with the memory, memcached & redis implementations for both sync & async we could consider moving this to master and I can take a crack at the mongodb implementation (I'm not sure anyone actually uses etcd so might not be worth the effort).

**Comment by Merinorus** [comment]
Thank you!
For the incrementation rework, while it's faster in case of DDoS attacks, it might be slower in normal use: when the user hits the limiter for the first time, it will need three memcached operations instead of two. Hence, the first implementation, maybe... But I guess it's better to favor IOPS when it's under pressure (DDoS, etc).

Sorry if it's not clear to me: is the current synchronous implementation of Redis and in-memory limiter okay for you? Or should I replicate the new Memcached version with the time-generated keys on these backbends as well?

**Comment by alisaifee** [comment]
> Sorry if it's not clear to me: is the current synchronous implementation of Redis and in-memory limiter okay for you? Or should I replicate the new Memcached version with the time-generated keys on these backbends as well?

My apologies for not clarifying, I think the current versions for in-memory & redis are fine (and in fact, I will probably use a similar approach for mongodb) since the atomicity of the operations (and the current/previous swap) can be managed on the (database) server side.

**Comment by alisaifee** [comment]
Just FYI: this is the WIP for mongodb: https://github.com/alisaifee/limits/commit/f0534392af0667fe6f653a4ae041ab6e9eef19bc

File: limits/aio/storage/memcached.py [hunk 1]
```diff
@@ -150,3 +229,91 @@ async def check(self) -> bool:
...
+            previous_key, current_key, expiry, get_current_ttl=False
+        )
+
+        weighted_count = previous_count * previous_ttl / expiry + current_count
+        if weighted_count + amount > limit:
+            return False
+        else:
+            # Hit, increase the current counter.
+            # If the counter doesn't exist yet, set twice the theorical expiry.
+            await self.incr(current_key, expiry + ceil(now % expiry), amount=amount)
```
Comment (alisaifee) [inline limits/aio/storage/memcached.py]: I think changing this to use `expiry * 2` would ensure the window stats are accurate and the failing test case is addressed.

Comment (Merinorus) [reply limits/aio/storage/memcached.py]: You're right. The delay should be set to expiry * 2, otherwise the counter might be reset too soon. I modified the sync implementation, I have to do the same with the async one.
However, the current failing test is not related to this. I think it's because the Memcached timer behavior is different from the others. Since I use a timestamp-based key, the reset time depends on the local machine timestamp (like the fixed window algorithm), whereas with Redis, it depends on the time of the first hit (like the moving window algorithm). I'm working on it.

Also, the reset time is not trivial:
- If only the current window counter is set (the easiest case), the reset time is the expiration of the current window,
- If the previous window is set, we can make it behave like the moving window by guessing the timestamp of the first hit, assuming the hits are evenly distributed on the previous window.

Edit: for the previous window, it's much easier to start at t=0s. A small burst might be allowed (if we time the end of the current window and if the previous window is empty), otherwise, it works well, especially with a low counter such as 1 request per second. 
For instance, with a 10s period, the previous window will be computed as follows:
- 1 hit : [0s]
- 2 hits: [0s, 5s]
- 3 hits: [0s, 3.33s, 6.66s]
- 4 hits: [0s, 2.5s, 5s, 7.5s]

Edit 2: for the concurrency problem with the in-memory async limiter, I think I should rewrite it. The timestamp-based key like with Memcached should do the trick.

Comment (alisaifee) [reply limits/aio/storage/memcached.py]: Thanks for sharing your findings! I've been trying to wrap my head around some of the edge cases (while working on the mongodb/etcd storage implementations). 

At the moment, I'm still seeing a few (flaky) test results:
- [mongodb](https://github.com/alisaifee/limits/tree/sliding-window-counter-mongodb) I've tried to emulate your implementation in redis 
- [etcd](https://github.com/alisaifee/limits/tree/sliding-window-etcd) Similar to your memcached implementation but without actually decrementing in case of a race condition allowing an increment to exceed the current window

Comment (Merinorus) [reply limits/aio/storage/memcached.py]: I did some more tests. We could allow exceeds in the current window. However, I cannot guarantee the upper limit of excess hits when the backend storage is under pressure. When testing locally, I did have 3-4 more hits than expected. Then I tried to add random delay on the `get` and `incr` operations (eg: `await asyncio.sleep(random.random() / 1000)`, and I could get not one, but more than a thousand more hits than expected. That's why I added the decrementation.

I didn't find a way to simulate a read/write delay properly in my tests.

Comment (alisaifee) [reply limits/aio/storage/memcached.py]: With the etcd & mongodb implementions, my naive solution was to fail the acquisition if the increment on the current window resulted in exceeding the limit and that passed the concurrency tests - however, both mongodb and etcd are not at all comparable in terms of performance with redis & memcached.

**Comment by Merinorus** [comment]
I struggle to understand the in-memory rate limiter.

I'm looking at the current in-memory limiter. With fixed window for instance, the following test is written:
```python
@pytest.mark.integration
class TestConcurrency:
    @all_storage
    def test_fixed_window(self, uri, args, fixture):
        storage = storage_from_string(uri, **args)
        limiter = limits.strategies.FixedWindowRateLimiter(storage)
        limit = RateLimitItemPerMinute(5)

        [limiter.hit(limit, uuid4().hex) for _ in range(50)]

        key = uuid4().hex
        hits = []

        def hit():
            time.sleep(random.random())
            if limiter.hit(limit, key):
                hits.append(None)

        threads = [threading.Thread(target=hit) for _ in range(50)]
        [t.start() for t in threads]
        [t.join() for t in threads]

        assert len(hits) == 5
```

It checks the behavior by launching multiple threads. But the fixed limiter just does increments on hits:

```python
class FixedWindowRateLimiter(RateLimiter):
    """
    Reference: :ref:`strategies:fixed window`
    """

    def hit(self, item: RateLimitItem, *identifiers: str, cost: int = 1) -> bool:
        """
        Consume the rate limit

        :param item: The rate limit item
        :param identifiers: variable list of strings to uniquely identify this
         instance of the limit
        :param cost: The cost of this hit, default 1
        """

        return (
            self.storage.incr(
                item.key_for(*identifiers),
                item.get_expiry(),
                elastic_expiry=False,
                amount=cost,
            )
            <= item.amount
        )
```

The incrementation function itself:

```python
    def incr(
        self, key: str, expiry: float, elastic_expiry: bool = False, amount: int = 1
    ) -> int:
        """
        increments the counter for a given rate limit key

        :param key: the key to increment
        :param expiry: amount in seconds for the key to expire in
        :param elastic_expiry: whether to keep extending the rate limit
         window every hit.
        :param amount: the number to increment by
        """
        self.get(key)
        self.__schedule_expiry()
        self.storage[key] += amount

        if elastic_expiry or self.storage[key] == amount:
            self.expirations[key] = time.time() + expiry

        return self.storage.get(key, 0)
```

Is that thread-safe? Shouldn't we acquire a lock whenever we want to increment the counter? The Python Counter class itself is not thread-safe.

Example:

```python
from collections import Counter
import threading

c = Counter()
lock = threading.Lock()

def test():
    for x in range(100000):
        with lock:  # <<--------  Remove this line and the test will fail
            c["key"] += 1

threads = [threading.Thread(target=test) for _ in range(10)]
for t in threads:
    t.start()

for t in threads:
    t.join()

assert c.get("key") == 1000000, c
```

Shouldn't we add a lock context for all in-memory incrementations? 

Edit: here is an attempt: https://github.com/Merinorus/limits/pull/2/commits/9c98defea238362f1f7e333eff871790263c01d8
I'm surprised this doesn't affect the benchmark.

**Comment by alisaifee** [comment]
> I struggle to understand the in-memory rate limiter.

This is very interesting. Without looking at the code I was certain the `self.lock` was being used for the fixed window operations but once I looked at it code again, I agree - there is nothing ensuring thread safety, and the code has been that way since the very first version of the library (~2014).

To make things a bit more confusing, the example you shared actually doesn't fail for me when I remove the lock. I wonder if the addition assignment on the counter key becomes 'mostly concurrent' on Cpython because of the way the GIL guards byte code execution?

Out of curiousity - which version / build of python are you using where you can reproduce the test failure with your example snippet? 

Having said all of that, I agree that we should be using a lock for the increment operator. Your implementation using the defaultdict looks great!

**Comment by Merinorus** [comment]
> > I struggle to understand the in-memory rate limiter.
> 
> This is very interesting. Without looking at the code I was certain the `self.lock` was being used for the fixed window operations but once I looked at it code again, I agree - there is nothing ensuring thread safety, and the code has been that way since the very first version of the library (~2014).
> 
> To make things a bit more confusing, the example you shared actually doesn't fail for me when I remove the lock. I wonder if the addition assignment on the counter key becomes 'mostly concurrent' on Cpython because of the way the GIL guards byte code execution?
> 
> Out of curiousity - which version / build of python are you using where you can reproduce the test failure with your example snippet?
> 
> Having said all of that, I agree that we should be using a lock for the increment operator. Your implementation using the defaultdict looks great!

Interesting indeed! The code fails only with Python 3.8 and Python 3.9. Starting from Python 3.10, it doesn't need the lock anymore.

**Comment by Merinorus** [comment]
I've cleaned the code and squashed the commits, it's ready for review.
I've created a branch with your commits for MongoDB & etcd: https://github.com/Merinorus/limits/pull/3
I haven't taken the time yet to check the failing tests.

**Comment by alisaifee** [comment]
> Interesting indeed! The code fails only with Python 3.8 and Python 3.9. Starting from Python 3.10, it doesn't need the lock anymore.

To make matters worse, the issue shows up again on python 3.13 😆

**Comment by Merinorus** [comment]
> > Interesting indeed! The code fails only with Python 3.8 and Python 3.9. Starting from Python 3.10, it doesn't need the lock anymore.
> 
> To make matters worse, the issue shows up again on python 3.13 😆

Wow, well done, I've tested it from python 3.8 to 3.12.

**Comment by Merinorus** [comment]
Cleaned code and commits, to be continued here: https://github.com/alisaifee/limits/pull/248
